shumai/io/network.ts

```diff
@@ -19,27 +19,35 @@ export function encode(tensor: sm.Tensor): ArrayBuffer {
   return buf.buffer
 }
 
-export function decode(obj: Response | ArrayBuffer): sm.Tensor {
-  function impl(buf) {
-    const shape_len = new Int32Array(buf, 0, 2)[0]
-    const shape = new BigInt64Array(buf, 8 /* 8 byte offset mandated alignement */, shape_len)
-    const t = sm.tensor(new Float32Array(buf, 8 + 8 * shape_len))
-    return t.reshape(shape)
-  }
+export function decodeBuffer(buf: ArrayBuffer) {
+  const shape_len = new Int32Array(buf, 0, 2)[0]
+  const shape = new BigInt64Array(buf, 8 /* 8 byte offset mandated alignement */, shape_len)
+  const t = sm.tensor(new Float32Array(buf, 8 + 8 * shape_len))
+  return t.reshape(shape)
+}
 
-  if (obj.constructor === Response || obj.constructor === Request) {
-    return new Promise((resolve) => {
-      obj.arrayBuffer().then((buf) => {
-        resolve(impl(buf))
-      })
-    })
-  } else if (obj.constructor === ArrayBuffer) {
-    return impl(obj)
-  }
-  throw `Could not decode object: ${obj}`
+export async function decode(obj: Response | ArrayBuffer) {
+  return new Promise<sm.Tensor>((resolve, reject) => {
```

Comment:
I think because it's async now, you can do something like

```javascript
if (obj.constructor === Response || obj.constructor === Request) {
  const buf = await obj.arrayBuffer()
  return decodeBuffer(buf)
} else if (obj.constructor === ArrayBuffer) {
  return decodeBuffer(obj)
}
throw `Could not decode object: ${obj}`
```

Reply:
Fixed ^^